pkg/syncer/fullsync.go

```diff
@@ -1561,41 +1564,165 @@ func cleanupCnsMaps(k8sPVs map[string]string, vc string) {
 	}
 }
 
-// cleanupUnusedPVCsAndSnapshotsFromGuestCluster iterates over all PVCs and VolumeSnapshots and
-// filter ones with DeletionTimestamp set but CNS specific finalizer present. These finalizers are added to
-// PVCs and VolumeSnapshots from guest cluster during their creation.
-// If such objects found, remove the finalizer from those objects if corresponding guest cluster is not-running/deleted.
-// This code handles cases where namespace deletion causes guest cluster and its corresponding components
-// to be deleted but associated objects on supervisor remain stuck in Terminating state.
-func cleanupUnusedPVCsAndSnapshotsFromGuestCluster(ctx context.Context) {
+func RemoveCNSFinalizerFromPVCIfTKGClusterDeleted(ctx context.Context, k8sClient clientset.Interface,
+	pvc *v1.PersistentVolumeClaim, finalizerToRemove string, isNamespaceBeingDeleted bool) {
 	log := logger.GetLogger(ctx)
-	// Create K8S client and snapshotter client
-	k8sClient, err := k8s.NewClient(ctx)
+	// Create client to operator on Cluster object
+	restClientConfig, err := k8s.GetKubeConfig(ctx)
 	if err != nil {
-		log.Errorf("cleanupUnusedPVCsAndSnapshotsFromGuestCluster: Failed to get kubernetes client. Err: %+v", err)
+		msg := fmt.Sprintf("RemoveCNSFinalizerFromPVCIfTKGClusterDeleted: Failed to initialize rest clientconfig. "+
+			"Err: %+v", err)
+		log.Error(msg)
 		return
 	}
-	snapshotterClient, err := k8s.NewSnapshotterClient(ctx)
+	ccClient, err := k8s.NewClientForGroup(ctx, restClientConfig, ccV1beta1.GroupVersion.Group)
 	if err != nil {
-		log.Errorf("cleanupUnusedPVCsAndSnapshotsFromGuestCluster: failed to get snapshotterClient. Err: %v", err)
+		msg := fmt.Sprintf("RemoveCNSFinalizerFromPVCIfTKGClusterDeleted: Failed to get vmOperatorClient. "+
+			"Err: %+v", err)
+		log.Error(msg)
 		return
 	}
+	log.Debugf("RemoveCNSFinalizerFromPVCIfTKGClusterDeleted called for pvc %s/%s finalizer %d", pvc.Namespace, pvc.Name,
+		len(pvc.ObjectMeta.Finalizers))
+	okToRemoveFinalizer := false
+	if len(pvc.ObjectMeta.Finalizers) != 0 &&
+		(slices.Contains(pvc.ObjectMeta.Finalizers, finalizerToRemove)) {
+		if !isNamespaceBeingDeleted {
+			// Fetch guest cluster name from PVC label and check if that guest cluster,
+			// where associated PVC was created, is running or not.
+			// NOTE: This is to be executed only when not called on namespace deletion.
+			var tkcClusterName string
+			for key := range pvc.ObjectMeta.Labels {
+				if strings.Contains(key, "TKGService") {
+					tkcDetails := strings.Split(key, "/")
+					tkcClusterName = tkcDetails[0]
+					break
+				}
+			}
+			cc := &ccV1beta1.Cluster{}
+			err := ccClient.Get(ctx, client.ObjectKey{
+				Namespace: pvc.Namespace,
+				Name:      tkcClusterName,
```

Comment:
What if tkcClusterName is not found from the label?
Do we have case when this label is not present on the PVC?